File: javascript/packages/linter/src/rules/erb-no-interpolated-class-names.ts
import { isLiteralNode, isWhitespaceLiteral, splitLiteralsAtWhitespace, groupNodesByClass } from "@herb-tools/core"
import { IdentityPrinter } from "@herb-tools/printer"

import { ParserRule } from "../types.js"
import { AttributeVisitorMixin } from "./rule-utils.js"

import type { Node } from "@herb-tools/core"
import type { StaticAttributeDynamicValueParams } from "./rule-utils.js"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult } from "@herb-tools/core"

function groupToString(group: Node[]): string {
  return group.map(node => {
    if (isLiteralNode(node)) {
      return node.content
    }

    return IdentityPrinter.print(node, { ignoreErrors: true })
  }).join("")
}

class ERBNoInterpolatedClassNamesVisitor extends AttributeVisitorMixin {
  protected checkStaticAttributeDynamicValue({ attributeName, valueNodes, attributeNode }: StaticAttributeDynamicValueParams) {
    if (attributeName !== "class") return

    const splitNodes = splitLiteralsAtWhitespace(valueNodes)
    const groups = groupNodesByClass(splitNodes)

    for (const group of groups) {
      if (group.every(node => isWhitespaceLiteral(node))) continue

      const isInterpolated = group.some(node => !isLiteralNode(node))
      if (!isInterpolated) continue

      const hasAttachedLiteral = group.some(node => isLiteralNode(node) && node.content.trim())
      if (!hasAttachedLiteral) continue

      const className = groupToString(group)

      this.addOffense(
        `Avoid ERB interpolation inside class names: \`${className}\`. Use standalone ERB expressions that output complete class names instead.`,
        attributeNode.value!.location,
      )
    }
  }
}

export class ERBNoInterpolatedClassNamesRule extends ParserRule {
  static ruleName = "erb-no-interpolated-class-names"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoInterpolatedClassNamesVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_interpolated_class_names.rs
use crate::utils::attribute_visitor::{classify_attribute, AttributeKind};
use crate::utils::class_grouping::{group_nodes_by_class, split_literals_at_whitespace};
use crate::utils::tag_utils::get_attributes;

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

rule_visitor!(ERBNoInterpolatedClassNamesVisitor);
define_parser_rule!(
  ERBNoInterpolatedClassNamesRule,
  "erb-no-interpolated-class-names",
  Warning,
  ERBNoInterpolatedClassNamesVisitor
);

impl Visitor for ERBNoInterpolatedClassNamesVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    for attribute in get_attributes(node) {
      let value_nodes = match classify_attribute(attribute) {
        Some(AttributeKind::StaticNameDynamicValue {
          attribute_name, value_nodes, ..
        }) if attribute_name == "class" => value_nodes,

        _ => continue,
      };

      let value_location = match attribute.value.as_ref() {
        Some(value_node) => value_node.location.clone(),
        None => continue,
      };

      for group in group_nodes_by_class(split_literals_at_whitespace(value_nodes)) {
        if group.iter().all(|node| node.is_pure_whitespace()) {
          continue;
        }

        if !group.iter().any(|node| !node.is_literal()) {
          continue;
        }

        if !group.iter().any(|node| node.is_literal() && !node.content().trim().is_empty()) {
          continue;
        }

        let class_name: String = group.iter().map(|node| node.printed()).collect();

        self.add_offense(
          format!(
            "Avoid ERB interpolation inside class names: `{}`. Use standalone ERB expressions that output complete class names instead.",
            class_name
          ),
          value_location.clone(),
        );
      }
    }

    self.walk_html_open_tag_node(node);
  }
}
